### ISSUE TITLE
Allow for building/testing stand-alone HOMME on cori-gpu testbed

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/E3SM#4263]
We'd like to test using cori-gpu testbed (set of 20 nodes that have 2 skylake CPU's and 4 nvidia V100's) as this could help in preparation for Perlmutter (which will be different hardware for sure, but the SW stack may be similar).

I added 2 cmake config files in this branch `ndk/homme/add-cori-gpu-cmake`, but it also needs environment variables to work.

To build with cuda, here is an example:
```
setenv OMPI_CXX $REPO/components/homme/../../externals/kokkos/bin/nvcc_wrapper
setenv NVCC_WRAPPER_DEFAULT_COMPILER "pgc++"

cmake VERBOSE=1 -C $REPO/components/homme/cmake/machineFiles/cori-gpu-hpcsdk.cmake \
   -DQSIZE_D=10 -DPREQX_PLEV=128 -DPREQX_NP=4 \
   -DKokkos_ENABLE_SERIAL=TRUE \
   -DKokkos_ENABLE_OPENMP=FALSE \
   -DBUILD_HOMME_WITHOUT_PIOLIBRARY=TRUE \
   -DBUILD_HOMME_THETA_KOKKOS=TRUE \
   -DHOMME_ENABLE_COMPOSE=TRUE \
   -DENABLE_OPENMP=FALSE \
   -DENABLE_HORIZ_OPENMP=FALSE \
   -DENABLE_COLUMN_OPENMP=FALSE \
   -DBUILD_HOMME_SWEQX=FALSE \
   -DHOMME_PROJID=e3sm \
   -DBUILD_HOMME_SWIM=FALSE \
   -DHOMMEXX_GB_CONFIG=TRUE \
   -DCMAKE_BUILD_TYPE=Release \
   -DBASE_FFLAGS="" \
   -DCMAKE_Fortran_FLAGS_RELEASE="-gopt -O3" \
   -DPREQX_USE_ENERGY=FALSE $HOMME
```

and then without cuda:

```
cmake VERBOSE=1 -C $REPO/components/homme/cmake/machineFiles/cori-gpu-nocuda-hpcsdk.cmake \
   -DQSIZE_D=10 -DPREQX_PLEV=128 -DPREQX_NP=4 \
   -DKokkos_ENABLE_SERIAL=TRUE \
   -DKokkos_ENABLE_OPENMP=FALSE \
   -DBUILD_HOMME_WITHOUT_PIOLIBRARY=TRUE \
   -DBUILD_HOMME_THETA_KOKKOS=TRUE \
   -DHOMME_ENABLE_COMPOSE=TRUE \
   -DENABLE_OPENMP=FALSE \
   -DENABLE_HORIZ_OPENMP=FALSE \
   -DENABLE_COLUMN_OPENMP=FALSE \
   -DBUILD_HOMME_SWEQX=FALSE \
   -DHOMME_PROJID=e3sm \
   -DBUILD_HOMME_SWIM=FALSE \
   -DHOMMEXX_GB_CONFIG=TRUE \
   -DCMAKE_BUILD_TYPE=Release \
   -DCMAKE_Fortran_FLAGS_RELEASE="-gopt -O3" \
   -DPREQX_USE_ENERGY=FALSE $HOMME
```

Note I needed to downgrade the default optimizations for the fortran build as I hit internal compiler error. Default includes `-fas`t and I found that setting to `-gopt -O3` seems to avoid that issue.

I can build the `theta-l_kokkos` target in both cases. Without cuda, a small ne4 test problem works, however, I get a runtime error when trying to run on device.

**Comment by ndkeen** [comment]
With some instrumental help from @lukaszlacinski , I was able to confirm that I didn't actually have anything wrong with the configuration (though we made some improvements) and he then found/fixed the issue. Which there may be 2 different errors.

The first runtime error I saw on gpu:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaMemcpy(dst, src, n, cudaMemcpyDefault) error( cudaErrorIllegalAddress): an illegal memory access was encountered /global/cscratch1/sd/ndk/wacmy/ndk_homme_add-cori-gpu-cm
ake/externals/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:96
```

Which was fixed with commit message: "There was an implicit use of *this in compute_eta. The fix is the usual 'create-a-local-copy'."  Then we saw:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaEventSynchronize(cuda_instance->constantMemReusable) error( cudaErrorIllegalAddress): an illegal memory access was encountered /global/cscratch1/sd/ndk/wacmy/ndk_homme_add-cori-gpu-cmake/externals/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:346
```

which was also resolved.  Commit message: "In a KOKKOS_LAMBDA in HyperviscosityFunctorImpl.cpp, we were
accessing a class member (implicitly using the this pointer)."

And now the "tiny" (ne4) NGGPS benchmark (test) will run.

**Comment by ndkeen** [comment]
I hit some build errors when I tried a "make all" which I think will happen anytime we use `BUILD_HOMME_WITHOUT_PIOLIBRARY=TRUE`.

Since we do have NETCDF installed on cori-gpu, I decided to turn on PIO/scorpio.

Note that on this machine NCDIR/NFDIR are set with `module load hdf5-parallel/1.10.5 zlib netcdf-c netcdf-fortran`

The "make all" command now completes, but I haven't tried running tests, etc.
